### ISSUE TITLE
ReactAreaLights do not seem to work in a module bundler

### ISSUE THREAD

**Issue by drcmda** [mrdoob/three.js#17220]
##### Description of the problem

basically, area lights do not work with npm/node module resolution, see:

simplified: https://codesandbox.io/s/three-fibre-userender-test-rohv5

raw threejs arealight demo: https://codesandbox.io/s/dreamy-platform-3jpim

## why 

all files under examples/jsm pull from `../../../build/three.module.js` instead of `three`, which is arguably not correct. it will pull threejs two times in bundling systems that do not follow package.json:module (but "main" instead, which leads to build/three.js). it then creates two separate namespaces. uniforms.init() ends up writing to the wrong one. 

## solution

this is how a typical jsm file should look:

```jsx
import {
	ClampToEdgeWrapping,
	DataTexture,
	FloatType,
	LinearFilter,
	NearestFilter,
	RGBAFormat,
	ShaderLib,
	UVMapping,
	UniformsLib
} from "three";
```

that's how any other library works. i have never seen library that relies on a distro file - this has obvious pitfalls. i understand that this was made to make the html demos work, but imo this is a hack, and it now impacts production usage.

##### Three.js version

- [x] r107

##### Browser

- [x] Chrome

##### OS

- [x] macOS

**Comment by Mugen87** [comment]
> it looks like this file: https://github.com/mrdoob/three.js/blob/master/examples/jsm/lights/RectAreaLightUniformsLib.js is mutating a namespace.

I'm not I understand. Can you please explain in more detail what you mean?

 It seems to work fine in the official example:

https://threejs.org/examples/webgl_lights_rectarealight

**Comment by drcmda** [comment]
the example is broken, too: https://codesandbox.io/s/dreamy-platform-3jpim

as i said, it fetches modules under the three.module.js namespace and mutates them. but apps later fetch from "three", which i guess, becomes another namespace. the way three operates here is very unusual.

**Comment by drcmda** [comment]
PS. i have included a console.log right after the init() function, which writes ltc_1 and 2 into the shaderlib. both ltc_1 and 2 are empty, because three/Shaderlib and three.module/Shaderlib are two separate modules.

**Comment by Mugen87** [comment]
> i think this is actually worse, i hope i'm wrong but i believe that any project right now that relies on examples/jsm pulls two full threejs bundles.

I think this is not true. As you can see at the following repository, you can perform ES6 imports like so:
```js
61 lines (38 sloc)  1.21 KB
    
import {
	BoxBufferGeometry,
	Mesh,
	MeshBasicMaterial,
	PerspectiveCamera,
	Scene,
	WebGLRenderer
} from 'three';

import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls.js';
````
And the resulting bundle does not contain multiple version of `three.js`. https://github.com/Mugen87/three-jsm

It seems to me you are doing something wrong in your build or at codebox.

**Comment by Mugen87** [comment]
Related: https://discourse.threejs.org/t/jsm-modules-for-browser-double-import-with-typescript/8235

**Comment by drcmda** [comment]
this happens in all bundlers. create-react-app, webpack, codesandbox, parcel, etc. as i said, the reason is that three imports from a namespace that differs from "three", so these mutations end up in another space. that's how modules work. 

your example is sheer accident imo. it only works because rollup happens to pull from three.module via jsnext:main/module. if it would follow follow "main" or fetch three from source, you would have two three versions in that build output.

**Comment by Mugen87** [comment]
> your example works because rollup happens to pull from three.module via package.json:jsnext:main/module. if it would follow "main" or fetch three from source, you would have two threejs versions in the output and arealights would also not work.

I think rollup does it the correct way and fetches `three` from the intended location (like configured in `package.json`). I'm not familiar with the other bundlers but I suppose it's a matter of configuration.

> is there any reason why it does that, though?

`three.js` is not a pure `node.js` project. It's intended that you can directly load modules in HTML pages like in the official examples. If we would load modules from 'three', a dependency resolution is not possible.

**Comment by njarraud** [comment]
I do not understand the issue either. I use the node system located in examples/jsm and build my app with webpack. Everything works as intended and I do not have any problem of double threes import either.

**Comment by drcmda** [comment]
@njarraud  im more concerned with react area lights than doubling. you can see in both demos above that it doesn't work. webpack has the same problems once it pulls from main and not module, try it.

> three.js is not a pure node.js project. It's intended that you can directly load modules in HTML pages like in the official examples. If we would load modules from 'three', a dependency resolution is not possible.

this is a real problem. if this library is supposed to work with the tools that practically everyone uses (npm), it shouldn't rely on a volatile namespace. even for standard imports, this is not good practice. imo this is a hack, it works in that html file now, but it severely impacts the usefulness of jsm.

**Comment by drcmda** [comment]
btw here's proof for doubling:

html demo

![Screenshot 2019-08-10 at 00 08 21](https://user-images.githubusercontent.com/2223602/62811722-01bd7400-bb03-11e9-9626-e1af4634179e.png)

bundled

![Screenshot 2019-08-10 at 00 08 59](https://user-images.githubusercontent.com/2223602/62811732-126dea00-bb03-11e9-9bf1-167f0b99acbd.png)

as i suspected, it pulls three.js and three.module.js, and creates two namespaces. you can assume this has silently happened in each and every bundling system that doesn't follow the module field.

**Comment by njarraud** [comment]
> this is a real problem. i think this lib should work with the tools that everyone uses (npm).

The reason I like and use threejs is because it doesn't depend on any particular third party platform like Nodejs. That is a strength from my point of view.

I'll check if I have an issue with ReactAreaLights as well.

**Comment by drcmda** [comment]
> doesn't depend on any particular third party platform like Nodejs. That is a strength from my point of view.

That still does not make it OK to break it for the all the people that use it. Standard imports are still practically useless, as is evident by this very issue, since it couldn't just use "three". But even if they weren't, the root cause is a misunderstanding of how modules behave.

**Comment by njarraud** [comment]
> That still does not make it OK to break it for the all the people that use it.

It doesn't make it OK either to break it for people who don't use it. I leave it to @mrdoob and the other three.js guys as they are definitely more qualify than I am to answer you concerns.

**Comment by gkjohnson** [comment]
I'm using webpack with babel, react, three.js, and a variety of three.js example modules and I'm not seeing a any duplicate threejs bundles included. I set up two small example builds using rollup and parcel and both seemed to work just fine, as well.

> if it would follow "main" or fetch three from source, you would have two threejs versions in the output and arealights would also not work.

Can you describe the setup you're using such that the build process is using the build pointed to by package.json "main"? And by fetching from source do you mean pointing to the "src/Three.js" file? My experience is that the bundlers [prioritize the "module" field over "main"](https://webpack.js.org/configuration/resolve/#resolvemainfields) so I wouldn't expect that this would be a problem.

I expect that the solution won't be to immediately change all the example scripts over to import from "three" because that would break quite a few things -- maybe we can talk about solutions that afford both use cases instead.

**Comment by drcmda** [comment]
> It doesn't make it OK either to break it for people who don't use it

I think you're missing my point. Right now this is a hack. By correcting the hack you don't break anything. That the html demos don't work has little bearing, because standard module imports in the real world do not exist. Imports have to run through a build system to resolve packages, like "three". In a real world application you would put a folder into /src like node_modules or dependencies, and import from there - this can and should be done here.

@gkjohnson Here's the statement from Ives, the guy who made csb: https://twitter.com/0xca0a/status/1159954977672105984

Threejs has two exports, a commonjs module `build/three` and a esm module `build/three.module`, it is up to the bundler to decide which it pulls and both are permissable. If a library on the other hand imports from "three/build/three.module", that is not correct, because we're skipping resolution now. all bundlers that follow main will have two bundles, as you can see in the demos. this will also happen in webpack, just tell it to pull main and you can see it.

>  scripts over to import from "three" because that would break quite a few things

pulling from "three" is the only sane way of importing a lib, there is no alternative. a solution would be to put "three" in all jsm files, so that they become working modules again. then the html demos have to be compiled into shape.

**Comment by looeee** [comment]
I've bundled three.js with webpack on a couple of projects and never noticed this happening. This sounds like it may be an issue with codesandbox. 

@drcmda, could you create a standalone package in a zip file using webpack or another bundler, so that we can run it locally and verify the issue?

**Comment by gkjohnson** [comment]
> It's up to the bundler to decide which it pulls and both module/main are permissable.
> all bundlers that follow main will have two bundles and all interop will break

I didn't expect this to be happening in practice -- all testing and experience indicates that module is being used over main. I see that at least CSB is using main, though. It would still be nice to know in what cases this happens with different bundlers.

I'll be the first to agree that the state of javascript imports is a bit of a mess and I don't think it's always clear what the "best" way to provide packages is right now. Because the examples reference a file they can be used directly in a browser through rawgit, unpkg, and just served with a static server for dev -- I feel that ease of use is something that's defined the library for awhile, now. I don't want to ignore the problem you've brought up but I do think it would be good to understand how common the issue is.

I also think it's most productive to discuss solutions to the problem, as well, so any thoughts you have in that vein would be appreciated. If the example modules are going to be switched over to use "three" then there will have to be some plan to keep the examples working and testable via rawgit.

**Comment by drcmda** [comment]
> I feel that ease of use is something that's defined the library for awhile, now. I don't want to ignore the problem you've brought up but I do think it would be good to understand how common the issue is.

That's my main concern. Is anyone here using standard imports? And how? You would not be able to use any package or npm depedency. I have never seen a single project on standard imports in the world. Imo the spec is merely theoretic at this point.

Npm on the other hands is used by the entire world. And here threejs has always thrown a fork in it. It was amazing to see that it finally moved away from script tags. But apparantly it still has a major problem with modules, as it does things nothing else does. As it stands, examples break the module standard, and that directly goes against what you said, ease of use.

**Comment by looeee** [comment]
> Is anyone here using browser imports? And how? 

I also use them for development since it enables me to avoid waiting for the bundler every time I make a change. 

I also use them for the examples on discoverthreejs.com (all of which run on codesandbox btw).

> at the expense of things like codesandbox or any other bundler that picks main, which are actually useful.

You haven't demonstrated a bundler aside from codesandbox that does this. 

Until you do that, I'm inclined to think that this is a codesandbox issue.

**Comment by drcmda** [comment]
It's not a bundler issue, this is how modules work. Three is published with two exports, main and module, see here: https://github.com/mrdoob/three.js/blob/dev/package.json#L5-L8 Both are two separate namespaces. Yet in the examples it clings to a single namespace, in no possible scenario is this correct. Even for those that are out there using draft imports - this is wrong.

Ives stated why he's doing it (in order to not transpile inside the browser) and there can be various reasons, but the issue won't go away. Examples is broken for anyone that uses the cjs module export right now.

**Comment by gkjohnson** [comment]
@drcmda I think you're stance on this is clear -- I don't feel repeating your point is going to help move this issue anywhere and ignoring requests for more information, examples, or suggestions isn't making it easy to collaborate, either.

#16920 uses rollup to convert all the jsm files into the UMD equivelant in the `/js` folder which includes adding `require('three')` statements to import three.js. As it is I think merging that PR and updating the docs a little with where to import examples when using bundlers is enough to handle this for the moment.

**Comment by looeee** [comment]
> Ives stated why he's doing it (in order to not transpile inside the browser) 

That may make sense for codesandbox, but it's not a standard or common use case. Ives also says that he's seen this causing problems with other libraries aside from three.js, and that they may change it on their side anyway.
 
This is the 'standard' usage of three.js via npm, and it works:

* install three from npm in a local dev environment
* es6 import, say, renderer, camera, and controls from the npm package
* Bundle with your preferred bundler: rollup, webpack, parcel. 

As far as I can see you want us to support using the cjs `build\three.js` script along with es6 example files. That's not a standard use case, and I don't think it's something we need to support, especially if doing so means we'd have to refactor all the examples. 

Unless you can do one of these :

* demonstrate a compelling and common example of this causing problems, aside from codesandbox
* come up with a solution that allows us to continue using browser imports in the examples

then I think we should close this as "won't fix".

**Comment by looeee** [comment]
This is the 'standard' usage of three.js via npm, and it works fine:

* install three from npm in a local dev environment
* es6 import, say, renderer, camera, and controls from the npm package
* Bundle with your preferred bundler: rollup, webpack, parcel. 

What's happening with codesandbox is that they doing something nonstandard to save time. When they encounter this:

``` js
import * as THREE from "three";
import { OrbitControls } from "three/examples/jsm/controls/OrbitControls.js";
```

For the first import, they are ignoring the es6 'module' field and sticking with the 'main' cjs field, to save time. 

But then they come to the `OrbitControls` import which specifies a single es6 file. There's no room for interpretation there and they have to use the es6 import... So now they are mixing cjs and es6 versions of three.js and you get two copies of the library in the bundle.

The issue here is that codesandbox is treating an es6 import as a cjs import. That's not correct and even though it works it has side effects like this. 

However, I don't think this is a common issue, and it's not something we need to worry about. Codesandbox can fix it on their side if they choose, or they can keep doing this for faster builds at the expense of bigger bundles. 

#16920 will also provide a solution for most people who want to use cjs imports but I don't think it will fix the codesandbox problem. 

For everyone else doing things the standard way, they'll never come across this issue.

**Comment by drcmda** [comment]
>  I don't feel repeating your point is going to help move this issue anywhere

I must say i feel kinda helpless. You are requesting more examples, but the root issue is ignored.

> What's happening with codesandbox is that they doing something nonstandard to save time. When they encounter this:

No. They don't do anything non standard. They just pick the commonjs module, which threejs officially exports. I have linked you package.json, you see these two modules in there.

**Comment by looeee** [comment]
> You are requesting more examples, but the root issue is ignored

I'm requesting any example of this happening anywhere except on codesandbox, a request that you have refused to acknowledge so far.

> No. They don't do anything non standard. They just pick the commonjs module, which threejs officially exports. 

This is a commonjs import, and should target 'main':

``` js
const THREE = require( 'three' );
```

This is an es6 import and should target 'module':

```js 
import * as THREE from 'three';
```

Doing otherwise, as codesandbox is doing, is not standard. In fact, even though some compilers may allow it, it's a violation of the es6 spec according to [this post on stackoverflow](https://stackoverflow.com/questions/29596714/new-es6-syntax-for-importing-commonjs-amd-modules-i-e-import-foo-require): 

> The TypeScript compiler at some point started allowing import * as foo from 'legacy-module-foo' to get the default import of a legacy module in certain circumstances. This is a violation of the ES6 specification (§15.2.1.16, __“The value "*" indicates that the import request is for the target module’s namespace object.”__).

**Comment by drcmda** [comment]
> I'm requesting any example of this happening anywhere except on codesandbox, a request that you have refused to acknowledge so far.

You still do not understand the root issue. I am sorry but you just do not. The bundle has nothing whatsoever to do with anything.

If you fetch the commonjs module, with whatever system or means at your disposal then that creates a namespace. This namespace will always differ from the namespace that jsm/examples opens up, because it targets another module.

In other words, commonJS and jsm/examples are incompatible.

here's another example in webpack/create-react-app: https://github.com/drcmda/three-cjs-bug/tree/master/src

it picks the commonjs export via

```jsx
import * as THREE from "three/build/three"
```

**Comment by Mugen87** [comment]
Well, what you are doing in your repository is obviously wrong. This is no demonstration of a bug. Yes, you can't use the UMD build `three.js` with JSM modules since they rely on `three.module.js`. However, this should be no problem if you have correct imports and a bundler works according to the standard.

However, I would appreciate if you recognize @looeee mentioned in his previous post. It seems to me that codesandbox needs a fix. Not `three.js`.

**Comment by looeee** [comment]
Your new example is incorrect since you are mixing cjs and es6 modules from a single npm package.

```js
// This is a cjs module
import * as THREE from "three/build/three"
```

``` js
// This is an es6 module
import { OrbitControls } from "three/examples/jsm/controls/OrbitControls.js"

```
 Just change that to `import * as THREE from "three"`, as it should be, and it will work fine. 

It seems like webpack is smart enough to handle cjs imports even though it's against the spec. Great. But if you really want to use the cjs version of three then you'll need to specifically request the cjs versions of the examples too, which you'll be able to do once #16920 lands. 

To be clear, any bundler that encounter  `'import * as X from 'package'` in an npm package should use the 'module' field if present. 

If they do otherwise, as codesandbox is doing, then they are incorrect and should expect problems like this one.

**Comment by gkjohnson** [comment]
@drcmda
> I must say i feel kinda helpless. You are requesting more examples, but the root issue is ignored. Not a single person so far has acknowledged the fact that threejs offers two modules.

Sorry I didn't mean to come across that way -- I understand the issue you're describing and acknowledge that there can be an issue in some cases. It is possible to configure these bundling systems to pull from "main" instead of "module" but it seems like a nonstandard workflow.

I'll be quick to admit that imports are a mess right now and I think we're just trying our best to make it work as cleanly for us as possible. Hopefully #16920 helps with that.

@looeee
> It seems like webpack is smart enough to handle cjs imports even though it's against the spec. Great. But if you really want to use the cjs version of three then you'll need to specifically request the cjs versions of the examples too, which you'll be able to do once #16920 lands.

I'm sure this is what you meant but just to be clear the cjs examples should be able to be used universally when bundling three.js in an application with any of these tools. `require('three')` should resolve to the "module" field when importing three.js so it can be intermixed with `import * as THREE from 'three'`.

**Comment by looeee** [comment]
> I'm sure this is what you meant but just to be clear the cjs examples should be able to be used universally when bundling three.js in an application with any of these tools. 

Well, yes, if you tell a tool to import a cjs file then it should do that. It's also fine for tools to import cjs using `import * as` - it seems like that's violation of the spec, but it's very convenient and common.

But if there is ambiguity, then a tool should always default to the es6 module when using  `import` and always default to cjs when using `require`. For example, `import * as THREE from 'three'` is ambiguous since it's not telling the bundler exactly what file to use. But it's using `import` so it should always choose `three.module.js` if available. If you want cjs then you should have to specify it using `import * as THREE from "three/build/three"`.

If a bundler doesn't do this then you have a situation where this:

``` js 
import * as THREE from "three";
import { OrbitControls } from "three/examples/jsm/controls/OrbitControls.js";
```

which is correct code, leads to wrong results. And that's just crazy and frustrating. 

The only thing I'd say is 'wrong' here is when a person or bundler uses cjs and es6 from the same npm package. Of course that's still a relative statement and if it works in your project then great. But doing that is what's causing all the confusion here and since codesandbox is not a private app then it seems like they should be more responsible here and at least warn people they are `import`ing cjs even though an es6 module is available.

**Comment by looeee** [comment]
Closing this issue since it's a problem with codesandbox, not three.js

**Comment by mrdoob** [comment]
@drcmda Sorry for the delay. Just to make sure I understand...

Are you suggesting removing `"main: "build/three.js"` from `package.json`?

Or removing `"module": "build/three.module.js"` and setting `"main": "build/three.module.js"`?

**Comment by mrdoob** [comment]
And, another noob question...

Say that `three` resolves to `three/build/three.module.js` and the dev writes this:

```js
import * as THREE from "three";
import { OrbitControls } from "three/examples/jsm/controls/OrbitControls.js";
``` 

Are you saying that bundlers are not able to tell that the `three/build/three.module.js` that `three` resolves to and the `../../build/three.module.js` that `OrbitControls.js` import from are the same file?

**Comment by drcmda** [comment]
@mrdoob if commonjs is unsupported i would indeed remove it. there would only be one export: build/three.module.js and everything starts working again. i have since bumped into more and more of the same problem in similar ciurcumstances. the dracoloader is broken too for instance - all due to module mismatch.

> Are you saying that bundlers are not able to tell that the three/build/three.module.js that three resolves to and the ../../build/three.module.js that OrbitControls.js import from are the same file?

Exactly. They are two separate namespaces. There is no resolution happening if three/examples targets a specific module, or rather, the bundler has no chance of resolving it.

**Comment by mrdoob** [comment]
>> Are you saying that bundlers are not able to tell that the three/build/three.module.js that three resolves to and the ../../build/three.module.js that OrbitControls.js import from are the same file?
>
> Exactly. They are two separate namespaces. There is no resolution happening if three/examples targets a specific module (build/three.module), while the bundle goes through standard package.json resolution and picks up commonjs, it has no chance of resolving the two properly.

I think I was not clear. In the situation I'm proposing there is no commonjs export, only a single export `"main": "build/three.module.js"`.

Would a bundler not be able to tell that tell that the `three/build/three.module.js` that `three` resolves to and the `../../build/three.module.js` that `OrbitControls.js` import from are the same file?

**Comment by drcmda** [comment]
yes, that would work. now resolution and three/examples point to the same file and that's that, it will work. you could still distribute the commonjs file so people can still use it - but if you don't declare it in package json, we're good.

well, not a 100%, because the commonjs module in build would still be incompatible, but cjs is dying and that's a good compromise i think.

**Comment by mrdoob** [comment]
Okay.

And, in order to do that, is it just a matter of removing removing `"module": "build/three.module.js"` from `package.json` and changing `"main": "build/three.js"` to `"main": "build/three.module.js"`?

**Comment by mrdoob** [comment]
Can we also remove `jsnext:main`?

**Comment by drcmda** [comment]
i think at least module and main could point to the same thing. jsnext was obsolete, and if it's gone old bundlers would probably go towards main i guess. but any of these variations will work, main only, main and module or all three - pointing to build/three.module

**Comment by mrdoob** [comment]
Okay. I'll try things for next release.

On npm-land, the only thing that will break is for people that rely on the commonjs build, right?

They'll have to update their code to `require( 'three/build/three.js' )`?

**Comment by mrdoob** [comment]
@drcmda Many thanks for being so patient explaining all this 🙏

**Comment by drcmda** [comment]
i think so, but it is more likely they were using "three" via resolution anyway. the only thing that could happen now is that their bundler doesn't understand the import statement, but i think there's no bundler left anymore ... webpack 1 or 2 maybe. in that case they import from 'build/three' like you said. thank you so much, too! :-)

**Comment by gkjohnson** [comment]
@mrdoob 
> And, in order to do that, is it just a matter of removing "module": "build/three.module.js" from package.json and changing "main": "build/three.js" to "main": "build/three.module.js"?

I'm not sure what your plan is with #16920 but if you're going to remove support for resolving "three" to the commonjs build then it may not make sense support `require` in the `jsm` -> `js` rollup conversion. It might make more sense to just produce an "iife" file, which would simplify the output a bit and prevent the js examples from breaking when calling `require('three')` in node.

**Comment by mrdoob** [comment]
> I'm not sure what your plan is with #16920 but if you're going to remove support for resolving "three" to the commonjs build then it may not make sense support `require` in the `jsm` -> `js` rollup conversion.

Yeah, that's why I was holding off that PR until things become more clear. I was even considering removing `examples/js` altogether, but I feel it's way too early for that?

I want to break as few things as possible while keeping the library approachable for people not using bundlers. And also painless for people using bundlers.

Is it okay to think that the bundler users that rely in commonjs shouldn't have a hard time updating to jsm?

**Comment by gkjohnson** [comment]
> I want to break as few things as possible while keeping the library approachable for people not using bundlers. And also painless for people using bundlers.

I entirely agree and I would like to be able to propose more solutions but I'm having a hard time discussing this without any tangible, demonstrable examples of the problem with a bundler. It's not that I don't believe it can happen I just want to be more well informed. A small zip project demonstrating the loading problem when using DracoLoader, for example, would be fantastic. Without something like this I'm not sure how any of us can be expected to sympathetically discuss this.

Having said that I don't have a problem with changing "main" to point to `three.module.js` myself except it may complicate things when importing into node (see below). It's admittedly at odds with what the [npm docs on package.json](https://docs.npmjs.com/files/package.json#main) imply should be in that the field, though.

> I was even considering removing examples/js altogether, but I feel it's way too early for that?

My opinion is that it's too early for that but if it's going to happen maybe it's worth embedding a deprecation warning log in the js example files for a few releases? I'd want a chance to clean up some of my github projects that depend on links to them, for example.

One other use case this impacts that likely isn't common is importing three.js and examples into node. I do this for small scripts, tests, benchmarks, etc. Changing "main" will mean `require('three')` will no longer work (which could be fine) and removing the `js/examples` will mean there's no process for importing example classes, loaders, etc into node without a build process, which to me would be a pain. It would be nice to have a reasonable way to support this (#16920 could have been an option). Right now those scripts will look something like this:

```js
global.THREE = require( 'three' );
require( 'three/examples/js/controls/OrbitControls.js' );

console.log( THREE.OrbitControls ); // woo!
```

All that aside would we be able to get #16920 merged a little sooner if I change the output to just IIFE meaning no functionality would be added or removed compared to the current js example files? At least then the dev process for modules would cleaner while this is sorted out.

> Is it okay to think that the bundler users that rely in commonjs shouldn't have a hard time updating to jsm?

In my experience most bundlers seem to let you use import and require interchangeably though maybe babel has to be involved in some cases?

**Comment by drcmda** [comment]
> A small zip project demonstrating the loading problem when using DracoLoader, for example, would be fantastic. Without something like this I'm not sure how any of us can be expected to sympathetically discuss this.

I think there have been enough demos though. All the above highlight the same problem. There's a sandbox and a github repo. The sandbox picks cjs because it just works that way. The GH repo picks cjs because it chooses to do so. And the result is: two three versions are loaded by the browser, which must naturally clash because three mutates its own namespace. Every thing in examples that relies, in any way, on namespace mutation is thereby broken, because examples is primed to the three.module namespace.

cjs itself is an obsolete concept, node has long moved on and understands imports, all bundlers handle imports, require could be marked obsolete here, too. In a way it already is, as we can see, since it doesn't even work with three's own addons. So removing it at least from package.json would only be logical.

**Comment by gkjohnson** [comment]
> Node has moved on and understands imports

I don't think this is a fair claim quite yet. Modules have been behind an experimental flag for years, [have recently had their implementation changed](https://medium.com/@nodejs/announcing-a-new-experimental-modules-1be8d2d6c2ff) and might be released for real in October. If the node.js case is considered worth supporting at all then I think removing the cjs version outright might be a bit hasty. But it sounds like it's coming sooner than I thought.

> I think there have been enough demos though. All of the above highlight the problem. The sandbox picks cjs because it just works that way. The GH repo picks cjs because it can.

I agree that codesandbox illustrates the issue as we've discussed it but [your github demo](https://github.com/drcmda/three-cjs-bug/blob/master/src/index.js#L2) does not. Here's why.

> ```
> import * as THREE from "three/build/three"
> ```

This line _subverts modules resulution_ which is the core of this discussion and therefore does not illustrate the problem. None of the three.js documention or standard use of the library support or advocate for the use of this import statement. Likewise someone could download their own version of three.js and import that as well as an example from this project and it would break in the same way.

None of the proposed solutions solve this case. That repo would break in the same way even after changing the example modules to resolve from `three` because the bundlers will resolve `three/build/three` to a different file from `three`. The same thing will happen if you change the main field in package.json. Do you understand my confusion? You've presented something as a repro case but none of the solutions you're happy with actually solve it so it cannot really be considered a repo case of the issue.

I'm looking for a repro of a bundler setup that illustrates importing from `three` and an example file but results in two versions of three.js included in the page. It was claimed previously that many projects may be inadvertently including two versions of three.js in the page if they use examples. How can this happen on accident? What would the bundler config look like?

If we're actually just talking about the codesandbox case then that's fine and @mrdoob's proposed solution will solve it just fine.

**Comment by drcmda** [comment]
> I don't think this is a fair claim quite yet. Modules have been behind an experimental flag for years

I am not talking about some draft spec here. The es spec does not matter, it does not exist, it should not be used. In the real world we use node resolution and everything has moved on from cjs, all bundlers and node. If there is someone that still relies on cjs - it is broken anyway, they can't, as i have demonstrated, use addons. To at least fix their three runtime, next version they would simply require("three/build/three") instead of "three" and that's that. But for 99.9% of all the other cases it work work, b/c it would only expose a single namespace.

> This line subverts modules resolution which is the core of this discussion and therefore does not illustrate the problem.

That _is_ the point. three/examples subverts module resolution. See: https://github.com/mrdoob/three.js/blob/dev/examples/jsm/lights/RectAreaLightUniformsLib.js#L16-L26

This example shows you that once you pick cjs it blows. You could also do this in the webpack config, with the same result. I hope this is clear by now, we are talking about a very benign and common thing, module namespaces. This has nothing to do with imports and bundlers. Three exposes two and clings to a single one in examples.

**Comment by gkjohnson** [comment]
Okay this sounds more like a philosphical discussion on the right / clean way to host and develop a javascript codebase independent of the fact that the current practice here is not breaking any real world projects with bundlers (aside from codesandbox). Am I understanding right?

**Comment by adrian-delgado** [comment]
If each jsm example would be its own package, then they would import from `'three'` directly and treat three as a peerDependency. If they are in the same package as `three` (but as subfolders), then it just happens that they can import either from `'three'` or from `'path/to/three.module.js'` (but not in all cases, like the codesandbox example).

This does seem to be more related to how es imports are meant to be treated. If `'some-package'` and `'some-package/sub-path'` are treated as independent modules, then the [real fix](https://github.com/mrdoob/three.js/issues/17220#issuecomment-529079066) proposed by @drcmda should be considered.

**Comment by drcmda** [comment]
> the current practice here is not breaking any real world projects with bundlers (aside from codesandbox)

it breaks commonjs, not codesandbox. codesandbox is broken because it uses commonjs. every project in the real world that is commonjs based cannot use three/examples. that is, it can, but half of it won't work.

> If each jsm example would be its own package, then they would import from 'three' directly and treat three as a peerDependency. If they are in the same package as three (but as subfolders), then it just happens that they can import either from 'three' or from 'path/to/three.module.js' (but not in all cases, like the codesandbox example).

im fine with either. but the moment three exposes two namespaces it won't work. three has to decide what it wants. it could fix examples, or remove one namespace from package json.

**Comment by mrdoob** [comment]
Moving the conversation to #17482. This thread got way too confusing.
